### PR TITLE
[12.0][MIG] stock_demand_estimate

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Standard migration.

### CODE REVIEW

File: stock_demand_estimate/wizards/stock_demand_estimate_wizard.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            ])
+            lines = []
+            for product in sheet.product_ids:
+                for _range in ranges:
+                    estimate = estimates.filtered(
+                        lambda x: (x.date_range_id == _range and
+                                   x.product_id == product)
+                    )
+                    if estimate:
+                        uom_id = estimate[0].product_uom.id
```
Comment (rousseldenis) [inline stock_demand_estimate/wizards/stock_demand_estimate_wizard.py]: @mreficent Please prefer use of :
```suggestion
                        from odoo.fields import first
                        uom_id = first(estimate).product_uom.id
```

File: stock_demand_estimate/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2016 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Stock Demand Estimate",
+    "summary": "Allows to create demand estimates.",
+    "version": "12.0.1.0.0",
+    "author": "Eficent, "
+              "Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-warehouse",
```
Comment (rousseldenis) [inline stock_demand_estimate/__manifest__.py:10]: Following https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py, you can add maintainers and development_status keys.

**Review by rousseldenis — APPROVED** [review]
Code review. Minor comments

File: stock_demand_estimate/wizards/stock_demand_estimate_wizard.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply stock_demand_estimate/wizards/stock_demand_estimate_wizard.py]: Just curious, what is the benefit of this change?

Comment (rousseldenis) [reply stock_demand_estimate/wizards/stock_demand_estimate_wizard.py]: It's not recommended to use indexes on recordset. Benefits of recordsets are kept. See https://github.com/odoo/odoo/blob/12.0/odoo/fields.py#L48

File: stock_demand_estimate/__manifest__.py (hunk 2 above)
Comment (MiquelRForgeFlow) [reply stock_demand_estimate/__manifest__.py:10]: But is optional

Comment (tbaden) [reply stock_demand_estimate/__manifest__.py:10]: installable attribute would be nice to have

File: stock_demand_estimate/models/stock_demand_estimate.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        digits=dp.get_precision('Product Unit of Measure'),
+    )
+    product_qty = fields.Float(
+        'Real Quantity',
+        compute='_compute_product_quantity',
+        inverse='_inverse_product_quantity',
+        digits=0,
+        store=True,
+        help='Quantity in the default UoM of the product',
+        readonly=True,
```
Comment (tbaden) [inline stock_demand_estimate/models/stock_demand_estimate.py]: compute does not need a readonly flag. compute is already readonly.

File: stock_demand_estimate/security/stock_security.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo noupdate="1">
+
+        <record model="ir.rule" id="stock_demand_estimate_comp_rule">
```
Comment (tbaden) [inline stock_demand_estimate/security/stock_security.xml]: >When declaring a record in XML:
Place id attribute before model

please set id as first attribute

File: stock_demand_estimate/views/date_range.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<odoo>
+
+    <menuitem
+        id="menuitem_date_range"
```
Comment (tbaden) [inline stock_demand_estimate/views/date_range.xml]: > Use the following patterns, where <model_name> is the name of the model that the menu, view, etc. belongs to (e.g. for a res.users form view, the name would be res_users_view_form):
> 
> For a menu: <model_name>_menu

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 6]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0"?>
+<odoo>
+
+    <record model="ir.ui.view"
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: > When declaring a record in XML:
> Place id attribute before model

> For a view: <model_name>_view_<view_type>, where view_type is kanban, form, tree, search, ...

please set the id correctly
stock_demand_estimate_view_tree

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                <field name="location_id"/>
+                <field name="product_uom_qty"/>
+                <field name="product_uom"/>
+                <field name="product_qty"/>
+                <field name="daily_qty"/>
+            </tree>
+        </field>
+    </record>
+
+    <record id="view_stock_demand_estimate_pivot" model="ir.ui.view">
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: > For a view: <model_name>_view_<view_type>, where view_type is kanban, form, tree, search, ...

please set the id correctly
stock_demand_estimate_view_pivot

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        <field name="arch" type="xml">
+            <pivot string="Stock Demand Estimate">
+                <field name="product_qty" type="measure"/>
+                <field name="product_id" type="row"/>
+                <field name="date_range_id" type="col"/>
+            </pivot>
+        </field>
+    </record>
+
+    <record id="view_stock_demand_estimate_search"
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: > When declaring a record in XML:
> Place id attribute before model

> For a view: <model_name>_view_<view_type>, where view_type is kanban, form, tree, search, ...

please set the id correctly
stock_demand_estimate_view_search

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        <field name="arch" type="xml">
+            <search string="Search Stock Demand Estimates">
+                <field name="date_range_id"/>
+                <field name="product_id"/>
+                <field name="location_id"/>
+            </search>
+        </field>
+    </record>
+
+    <record model="ir.actions.act_window"
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: please set id as first attribute

and

> For an action: the main action respects <model_name>_action. Others are suffixed with _<detail>, where detail is an underscore lowercase string explaining the action (should not be long). This is used only if multiple actions are declared for the model.

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            id="stock_demand_estimate_form_action">
+        <field name="name">Stock Demand Estimates</field>
+        <field name="type">ir.actions.act_window</field>
+        <field name="res_model">stock.demand.estimate</field>
+        <field name="view_type">form</field>
+        <field name="view_mode">tree,pivot</field>
+        <field name="search_view_id"
+               ref="view_stock_demand_estimate_search"/>
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: refactor as above mentioned

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    </record>
+
+    <menuitem id="menu_stock_demand_planning" name="Demand Planning"
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: > Use the following patterns, where <model_name> is the name of the model that the menu, view, etc. belongs to (e.g. for a res.users form view, the name would be res_users_view_form):
> 
> For a menu: <model_name>_menu

File: stock_demand_estimate/views/stock_demand_estimate_view.xml [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+              parent="stock.menu_stock_root" sequence="10"/>
+
+    <menuitem
+        id="menu_stock_demand_estimate"
```
Comment (tbaden) [inline stock_demand_estimate/views/stock_demand_estimate_view.xml]: > Use the following patterns, where <model_name> is the name of the model that the menu, view, etc. belongs to (e.g. for a res.users form view, the name would be res_users_view_form):
> 
> For a menu: <model_name>_menu

File: stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml [hunk 13]
```diff
@@ -0,0 +1,94 @@
+<?xml version="1.0"?>
+<odoo>
+
+        <record model="ir.ui.view"
```
Comment (tbaden) [inline stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml]: please set id as first item

and use id pattern:

> For a view: <model_name>_view_<view_type>, where view_type is kanban, form, tree, search, ...

File: stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+                            class="oe_highlight"/>
+                    <button class="oe_link"
+                            special="cancel"
+                            string="Cancel"/>
+                    </footer>
+                </form>
+          </field>
+        </record>
+
+        <record model="ir.ui.view"
```
Comment (tbaden) [inline stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml]: please set id as first item

and use id pattern:

> For a view: <model_name>_view_<view_type>, where view_type is kanban, form, tree, search, ...

File: stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        <act_window name="Create Stock Demand Estimates"
+            res_model="stock.demand.estimate.wizard"
+            src_model="stock.demand.estimate.sheet"
+            view_mode="form"
+            target="new"
+            key2="client_action_multi"
+            id="action_stock_demand_estimate_wizard"/>
+
+        <menuitem
+            id="menu_stock_demand_estimate_wizard"
```
Comment (tbaden) [inline stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml]: > Use the following patterns, where <model_name> is the name of the model that the menu, view, etc. belongs to (e.g. for a res.users form view, the name would be res_users_view_form):
> 
> For a menu: <model_name>_menu

File: stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+          </field>
+        </record>
+
+        <act_window name="Create Stock Demand Estimates"
+            res_model="stock.demand.estimate.wizard"
+            src_model="stock.demand.estimate.sheet"
+            view_mode="form"
+            target="new"
+            key2="client_action_multi"
+            id="action_stock_demand_estimate_wizard"/>
```
Comment (tbaden) [inline stock_demand_estimate/wizards/stock_demand_estimate_wizard_view.xml]: > For an action: the main action respects <model_name>_action. Others are suffixed with _<detail>, where detail is an underscore lowercase string explaining the action (should not be long). This is used only if multiple actions are declared for the model.

**Review by tbaden — CHANGES_REQUESTED** [review]
thanks for your work 👍 
i did a code review. here some changes:

**Review by gurneyalex — APPROVED** [review]

**Comment by rafaelbn** [comment]
Hi @mreficent could you please update branch to resolve conflict? if all is OK and green it could be merged!

**Comment by MiquelRForgeFlow** [comment]
@rafaelbn rebased
@tbaden comments attended

**Review by tbaden — APPROVED** [review]
code review: lgtm

**Comment by rousseldenis** [comment]
@mreficent Do you will squash two last commits or let as is ?

**Comment by MiquelRForgeFlow** [comment]
As you prefer. I wanted to not squash them in order to maintain your authorship of your contribution.

**Comment by rousseldenis** [comment]
> As you prefer. I wanted to not squash them in order to maintain your authorship of your contribution.

Too much honor :smile:

**Pull request merged**